ql/src/java/org/apache/hadoop/hive/ql/parse/type/RexNodeExprFactory.java

```diff
@@ -374,7 +373,7 @@ protected Object interpretConstantAsPrimitive(PrimitiveTypeInfo targetType, Obje
         HiveChar newValue = new HiveChar(constValue, length);
         HiveChar maxCharConst = new HiveChar(constValue, HiveChar.MAX_CHAR_LENGTH);
         if (maxCharConst.equals(newValue)) {
-          return makeHiveUnicodeString(Interpretation.CHAR, newValue.getValue());
+          return makeHiveUnicodeString(newValue.getValue());
```

Comment:
this will discard type distinction between char/varchar - but because `CHAR` is already padded at this point; it will work correctly!
awesome! :)